File: python/0055-jump-game.py
class Solution:
    # O(n) time | O(1) space
    def canJump(self, nums: List[int]) -> bool:
        goal = len(nums) - 1
        for idx in range(goal, -1, -1):
            if idx + nums[idx] >= goal:
                goal = idx
        return goal == 0

File: c++/0055-jump-game.cpp
class Solution {
public:
    // O(n) time, O(1) space - where n is the size of the input array
    bool canJump(vector<int>& nums) {
        int n = nums.size();
        int res = n - 1;
        for (int i = n - 2; i >= 0; --i) {
            if (nums[i] + i >= res) {
                res = i;
            }
        }
        return res == 0;
    }
};